From c40e246c767ce5a3a579ae0c1f08e11866a31979 Mon Sep 17 00:00:00 2001
From: "Srivatsa S. Bhat (VMware)" <srivatsa@csail.mit.edu>
Date: Mon, 24 Jan 2022 17:06:30 -0800
Subject: [PATCH] wireguard: Fix compilation error on linux-secure

linux-secure build fails at wireguard as shown below:

drivers/net/wireguard/queueing.c: In function 'wg_packet_queue_free':
drivers/net/wireguard/queueing.c:45:2: error: cast from function type
'void (*)(struct sk_buff *)' to non-matching type 'void (*)(void *)'
   45 |  ptr_ring_cleanup(&queue->ring, purge ?
   (void(*)(void*))kfree_skb : NULL);
         |  ^~~~~~~~~~~~~~~~
	 make[3]: *** [scripts/Makefile.build:280:
	 drivers/net/wireguard/queueing.o] Error 1
	 make[3]: *** Waiting for unfinished jobs....

Signed-off-by: Srivatsa S. Bhat (VMware) <srivatsa@csail.mit.edu>
---
 drivers/net/wireguard/queueing.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireguard/queueing.c b/drivers/net/wireguard/queueing.c
index 1de413b19e34..78fe1e7bf1f1 100644
--- a/drivers/net/wireguard/queueing.c
+++ b/drivers/net/wireguard/queueing.c
@@ -38,11 +38,16 @@ int wg_packet_queue_init(struct crypt_queue *queue, work_func_t function,
 	return 0;
 }
 
+static void wg_skb_cleanup(void *ptr)
+{
+	kfree_skb(ptr);
+}
+
 void wg_packet_queue_free(struct crypt_queue *queue, bool purge)
 {
 	free_percpu(queue->worker);
 	WARN_ON(!purge && !__ptr_ring_empty(&queue->ring));
-	ptr_ring_cleanup(&queue->ring, purge ? (void(*)(void*))kfree_skb : NULL);
+	ptr_ring_cleanup(&queue->ring, purge ? wg_skb_cleanup : NULL);
 }
 
 #define NEXT(skb) ((skb)->prev)
-- 
2.25.1

